### ISSUE TITLE
`DeepReadonly` *is excessively deep and possibly infinite* for a union with an array of itself

### ISSUE THREAD

**Issue by psmolinsky** [ts-essentials/ts-essentials#306]
If there is a union includes an array of the union itself, then the union wrapped with `DeepReadonly` does not compile with the latest TypeScript.

Sample code:

```typescript
type Foo = { value: string } | Foo[];

type ReadonlyFoo = DeepReadonly<Foo>;
```

It shows the compiler error:

```
Type instantiation is excessively deep and possibly infinite. ts(2589)
```

Suggested fix is simple: add an explicit branch for arrays to [DeepReadonly](https://github.com/ts-essentials/ts-essentials/blob/6b27ee4b7849d2a2670d80000a928cd20e3981b7/lib/types.ts#L154) before `T extends {}` put: `T extends AnyArray<infer U> ? readonly DeepReadonly<U>[] : ...`

**Comment by Beraliv** [comment]
Hey @psmolinsky!

Sorry for the late response, I was relocated to another country and still come back to normal schedule.

Thank you for the report!

Do you mind creating PR with tests and the fix?

I checked it for recursive object and array and it's not working only for arrays – https://tsplay.dev/NBkRVm

Sorry for the inconvenience in advance

**Comment by psmolinsky** [comment]
Hi @Beraliv, thanks for the guidance. I've created a PR. Please review.

**Comment by Beraliv** [comment]
It will be released in a separate PR, I will attach the link later

**Comment by Beraliv** [comment]
It will be released in ~1-2 weeks in https://github.com/ts-essentials/ts-essentials/pull/326

**Comment by Beraliv** [comment]
@psmolinsky thank you for your contribution again!

Released in 9.3.0 ✅
